internal/session.go

```diff
@@ -0,0 +1,421 @@
+	}
+
+	tasklistChan := GetSignalChannel(ctx, sessionID) // use sessionID as channel name
+	// Retry is only needed when creating new session and the error returned is NewCustomError(errTooManySessionsMsg)
+	retryPolicy := &RetryPolicy{
+		InitialInterval:    time.Second,
+		BackoffCoefficient: 1.1,
+		MaximumInterval:    time.Second * 10,
+		MaximumAttempts:    10,
+		NonRetriableErrorReasons: []string{
+			"cadenceInternal:Panic",
+			"cadenceInternal:Generic",
+			"cadenceInternal:Timeout START_TO_CLOSE",
+			"cadenceInternal:Timeout HEARTBEAT",
+		},
+	}
+	// Creation activity need an infinite StartToCloseTimeout.
+	scheduleToStartTimeout := time.Second * 10
+	startToCloseTimeout := time.Second*math.MaxInt32 - scheduleToStartTimeout
+	HeartbeatTimeout := time.Second * 20
+	ao := ActivityOptions{
+		TaskList:               creationTasklist,
+		ScheduleToStartTimeout: scheduleToStartTimeout,
+		StartToCloseTimeout:    startToCloseTimeout,
+		HeartbeatTimeout:       HeartbeatTimeout,
+	}
+	if retryable {
+		ao.RetryPolicy = retryPolicy
+	}
+	creationCtx := WithActivityOptions(ctx, ao)
+	creationFuture := ExecuteActivity(creationCtx, sessionCreationActivityName, sessionID)
+
+	var creationErr error
+	var tasklist string
+	s := NewSelector(ctx)
+	s.AddReceive(tasklistChan, func(c Channel, more bool) {
+		c.Receive(ctx, &tasklist)
+	})
+	s.AddFuture(creationFuture, func(f Future) {
+		// activity stoped before signal is received, must be creation timeout.
+		creationErr = f.Get(ctx, nil)
+	})
+	s.Select(ctx)
+
+	if creationErr != nil {
+		return nil, creationErr
+	}
+
+	sessionInfo := &SessionInfo{
+		SessionID:    sessionID,
+		tasklist:     tasklist,
+		sessionState: sessionStateOpen,
+	}
+
+	Go(ctx, func(ctx Context) {
+		if creationFuture.Get(ctx, nil) != nil {
+			sessionInfo.sessionState = sessionStateFailed
+		}
+	})
+
+	return setSessionInfo(ctx, sessionInfo), nil
+}
+
+func generateSessionID(ctx Context) (string, error) {
+	var sessionID string
+	err := SideEffect(ctx, func(ctx Context) interface{} {
+		return uuid.New()
+	}).Get(&sessionID)
+	return sessionID, err
+}
+
+func getCreationTasklist(base string) string {
+	return base + "__internal_session_creation"
+}
+
+func getResourceSpecificTasklist(identity, resourceID string) string {
+	return resourceID + "@" + identity
+}
+
+func sessionCreationActivity(ctx context.Context, sessionID string) error {
+	sessionEnv, ok := ctx.Value(sessionEnvironmentContextKey).(*sessionEnvironment)
+	if !ok {
+		panic("no session environment in context")
+	}
+	sessionEnv.Lock()
+	if sessionEnv.sessionTokenBucket.availableToken == 0 {
+		sessionEnv.Unlock()
+		return NewCustomError(errTooManySessionsMsg)
+	}
+
+	sessionEnv.sessionTokenBucket.availableToken--
+	doneChan := make(chan struct{})
+	sessionEnv.doneChanMap[sessionID] = doneChan
+	sessionEnv.Unlock()
+
+	defer func() {
+		sessionEnv.sessionTokenBucket.addToken()
+	}()
+
+	activityEnv := getActivityEnv(ctx)
+
+	if sessionEnv.testEnv == nil {
```

Comment:
I would remove testEnv from the sessionEnvironment and would create a version of Client that GetClient returns that delegates Signal to testEnv.

Reply:
I changed sessionEnvironment into an interface and provided two implementations. The test implementation will perform signal using testEnv.